jazkarta/shop/browser/checkout.py

```diff
@@ -24,13 +32,318 @@
+        return self.context.absolute_url()
+
+    def handle_submit(self):
+        amount = self.amount
+        response = self.request['x_response_code']
+        if response != '1':
+            # Authorize.Net SIM response codes
+            # 1 = Approved
+            # 2 = Declined
+            # 3 = Error
+            # 4 = Held for Review
+            if 'x_response_reason_text' in self.request.form:
+                self.error = self.request['x_response_reason_text']
+            else:
+                # this scenario unlikely as every response should provide us 
+                # with x_response_reason_text
+                self.error = ('There was an error with your transaction. '
+                              'Your payment has not been processed. '
+                              'Please contact us for assistance.')
+        if self.error:
+            return
+
+        userid = get_current_userid()
+        contact_info = PersistentMapping()
+        if 'x_first_name' in self.request.form:
+            contact_info['first_name'] = self.request.form['x_first_name']
+        if 'x_last_name' in self.request.form:
+            contact_info['last_name'] = self.request.form['x_last_name']
+        if 'x_email' in self.request.form:
+            contact_info['email'] = self.request.form['x_email']
+        if 'x_phone' in self.request.form:
+            contact_info['phone'] = self.request.form['x_phone']
+        if 'x_address' in self.request.form:
+            contact_info['address'] = self.request.form['x_address']
+        if 'x_city' in self.request.form:
+            contact_info['city'] = self.request.form['x_city']
+        if 'x_state' in self.request.form:
+            contact_info['state'] = self.request.form['x_state']
+        if 'x_zip' in self.request.form:
+            contact_info['zip'] = self.request.form['x_zip']
+        if 'x_country' in self.request.form:
+            contact_info['country'] = self.request.form['x_country']
+
+        method = 'Online Payment'
+        charge_result = {'success': True}
+
+        if not self.error:
+            try:
+                self.store_order(method, charge_result, userid, contact_info)
+                self.clear_cart()
+            except ConflictError:
+                raise Exception(
+                    'Failed to store results of payment after multiple '
+                    'retries. Please contact us for assistance.')
+
+    # This code runs after the payment is processed
+    # to update various data in the ZODB.
+    @run_in_transaction(retries=10)
+    def store_order(self, method, charge_result, userid, contact_info):
+        """Store various data following payment processing.
+
+        This code runs in a separate transaction (with up to 10 retries)
+        to make sure that ConflictErrors do not cause the Zope publisher
+        to retry the request which performs payment processing.
+
+        # ^ is the above still true when using Authorize.Net SIM??
```

Comment:
Still true. Probably less crucial because the normal transaction will be shorter (since it doesn't start until Auth.net calls us back, instead of starting before we call Stripe) but there's no harm in setting it up to try multiple times.